### PR TITLE
Fix: Setting DNI mandatory after address creation launches an exception during checkout

### LINKED ISSUE #31668: Setting DNI mandatory after address creation launches an exception during checkout
### Prerequisites

- [x] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

If for a country there is the mandatory CF field and try to make an order (in the front-office) with an address that has the DNI empty field, the system shows an error that I attach.
Screenshot 2023-03-07 at 10-22-40 Screenshot

### Expected behavior

Error/Alert message inviting the customer to complete all the mandatory fields in the address

### Steps to reproduce

1. Add a product to the cart (no account)
2. Go in the checkout and create an account
3. Add an address, select Italy as a country
4. Save the address (don't complete the checkout, very important)
5. As an administrator (BO), go to back office, International > Country, edit Italy and enable the DNI field as mandatory
6. As the new customer (FO) try to complete the checkout

The problem occurs due to the modification to the **restorePersistedData** method of the Controller **controllers/front/Ordercontroller.php**
Until version 1.7.8.7 of PS the education $ This-> Cartchecksum-> GenerateChecksum ($ cart) Venifa performed only if there were no non-valid addresses, while in 1.7.8.8 this is always performed and therefore one is launched "non -managed" exception

**PS 1.7.8.7**
```

    protected function restorePersistedData(CheckoutProcess $process)
    {
        $cart = $this->context->cart;
        $customer = $this->context->customer;
        $rawData = Db::getInstance()->getValue(
            'SELECT checkout_session_data FROM ' . _DB_PREFIX_ . 'cart WHERE id_cart = ' . (int) $cart->id
        );
        $data = json_decode($rawData, true);
        if (!is_array($data)) {
            $data = [];
        }

        $addressValidator = new AddressValidator();
        $invalidAddressIds = $addressValidator->validateCartAddresses($cart);

        // Build the currently selected address' warning message (if relevant)
        if (!$customer->isGuest() && !empty($invalidAddressIds)) {
            $this->checkoutWarning['address'] = [
                'id_address' => (int) reset($invalidAddressIds),
                'exception' => $this->trans(
                    'Your address is incomplete, please update it.',
                    [],
                    'Shop.Notifications.Error'
                ),
            ];

            $checksum = null;
        } else {
            $checksum = $this->cartChecksum->generateChecksum($cart);
        }

        // Prevent check for guests
        if ($customer->id) {
            // Prepare all other addresses' warning messages (if relevant).
            // These messages are displayed when changing the selected address.
            $allInvalidAddressIds = $addressValidator->validateCustomerAddresses($customer, $this->context->language);
            $this->checkoutWarning['invalid_addresses'] = $allInvalidAddressIds;
        }

        if (isset($data['checksum']) && $data['checksum'] === $checksum) {
            $process->restorePersistedData($data);
        }
    }

```

**PS 1.7.8.8**

```

    protected function restorePersistedData(CheckoutProcess $process)
    {
        $cart = $this->context->cart;
        $customer = $this->context->customer;
        $rawData = Db::getInstance()->getValue(
            'SELECT checkout_session_data FROM ' . _DB_PREFIX_ . 'cart WHERE id_cart = ' . (int) $cart->id
        );
        $data = json_decode($rawData, true);
        if (!is_array($data)) {
            $data = [];
        }

        $addressValidator = new AddressValidator();
        $invalidAddressIds = $addressValidator->validateCartAddresses($cart);

        // Build the currently selected address' warning message (if relevant)
        if (!$customer->isGuest() && !empty($invalidAddressIds)) {
            $this->checkoutWarning['address'] = [
                'id_address' => (int) reset($invalidAddressIds),
                'exception' => $this->trans(
                    'Your address is incomplete, please update it.',
                    [],
                    'Shop.Notifications.Error'
                ),
            ];
        }

        // Prevent check for guests
        if ($customer->id) {
            // Prepare all other addresses' warning messages (if relevant).
            // These messages are displayed when changing the selected address.
            $allInvalidAddressIds = $addressValidator->validateCustomerAddresses($customer, $this->context->language);
            $this->checkoutWarning['invalid_addresses'] = $allInvalidAddressIds;
        }

        if (isset($data['checksum']) && $data['checksum'] === $this->cartChecksum->generateChecksum($cart)) {
            $process->restorePersistedData($data);
        }
    }

```

### PrestaShop version(s) where the bug happened

1.7.8.8, 8.0.x, 8.1.0

### PHP version(s) where the bug happened

7.4, 8.0, 8.1

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by Codencode**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           |  8.1.x
| Description?      | If for a country there is the mandatory CF field and try to make an order (in the front-office) with an address that has the DNI empty field, the system shows an error that I attach.
| Type?             | bug fix 
| Category?         | FO
| BC breaks?        | no
| Deprecations?     |  no
| Fixed ticket?     | Fixes #31668
| Sponsor company   | Codencode snc

### CODE REVIEW

**Comment by Hlavtox** [comment]
I think that this is still not a good solution. With this, you are returning a simmilar checksum for all carts where there is no address valid.

As I said in your original PR:

> I think you should implement this check in AddressChecksumCore::generateChecksum and return sha1('No address set'); like it does if (!$address->id).

**Comment by Codencode** [comment]
> I think that this is still not a good solution. With this, you are returning a simmilar checksum for all carts where there is no address valid.
> 
> As I said in your original PR:
> 
> > I think you should implement this check in AddressChecksumCore::generateChecksum and return sha1('No address set'); like it does if (!$address->id).


You mean it's better to replace
`return sha1('No address set');`

with something like this?
` return sha1('Invalid address set');`

**Comment by Hlavtox** [comment]
@Codencode No, I would do it here - https://github.com/PrestaShop/PrestaShop/blob/develop/classes/AddressChecksumCore.php

But let's see what others think. Ping @kpodemski

**Comment by boherm** [comment]
> @Codencode No, I would do it here - https://github.com/PrestaShop/PrestaShop/blob/develop/classes/AddressChecksumCore.php
> 
> But let's see what others think. Ping @kpodemski

I agree! It can be done here.

**Comment by Codencode** [comment]
> Invalid address set

@Hlavtox @boherm 
So you would do it like this?
```
class AddressChecksumCore implements ChecksumInterface
{
    public const SEPARATOR = '_';

    /**
     * Generate a checksum.
     *
     * @param Address $address
     *
     * @return string SHA1 checksum for the Address
     */
    public function generateChecksum($address)
    {
        if (!$address->id) {
            return sha1('No address set');
        }

        // additional control *******************************************************
        if (Country::isNeedDniByCountryId($address->id_country) && !$address->dni) {
            return sha1('Dni is required'); 
        }
        //***************************************************************************

        $uniqId = '';
        $fields = $address->getFields();
        foreach ($fields as $name => $value) {
            $uniqId .= $value . self::SEPARATOR;
        }
        $uniqId = rtrim($uniqId, self::SEPARATOR);

        return sha1($uniqId);
    }
}
```

**Comment by boherm** [comment]
@Codencode, it can be I think! 👍 
But instead of `Country::isNeedDniByCountryId` I would rather write `Address::dniRequired` to avoid use another ObjectModel than Address.

**Comment by Hlavtox** [comment]
@Codencode Yeah that is much nicer dude. See, we still get a proper cart checksum if you do it like this. :-)

And @boherm also has a nice suggestion. There is a function like this.

```php
if (Address::dniRequired($address->id_country) && !$address->dni) {
    return sha1('Dni is required'); 
}
```

**Comment by Codencode** [comment]
@boherm good idea!
@Hlavtox thank you for your suggestion!

Should I delete this PR and create a new one?
Sorry but I'm still unfamiliar with Github

**Comment by boherm** [comment]
> Should I delete this PR and create a new one? Sorry but I'm still unfamiliar with Github

No need to create a new PR, you can add new commit with your new diff and rebase interactively your branch to remove your olders 2 commits, than push force your branch! With this method, after merging we only have one commit in PS history for your PR. :-)

**Comment by Codencode** [comment]
> No need to create a new PR, you can add new commit with your new diff and rebase interactively your branch to remove your olders 2 commits, than push force your branch! With this method, after merging we only have one commit in PS history for your PR. :-)

I made the change, but I don't know if it's okay because I don't understand the "rebase" argument, I "reverted" the changes that are no longer needed.

I await feedback.

Thank you!

**Comment by boherm** [comment]
Thanks for your changes @Codencode!
But I think we can have a git history cleaner, in fact only your last commit (465fcd1...) need to be in your PR! :D 
When you revert a commit, git create another commit in your history.

To clean your commits history, you can rewrite it like explain in https://www.atlassian.com/git/tutorials/rewriting-history#git-rebase-i. It could be cool if you do so, to avoid to add 7 commits into PrestaShop main history! :D

**Comment by Codencode** [comment]
> Thanks for your changes @Codencode! But I think we can have a git history cleaner, in fact only your last commit ([465fcd1](https://github.com/PrestaShop/PrestaShop/commit/465fcd1bdc3e413b537dd086b5195dd2f69cab99)...) need to be in your PR! :D When you revert a commit, git create another commit in your history.
> 
> To clean your commits history, you can rewrite it like explain in https://www.atlassian.com/git/tutorials/rewriting-history#git-rebase-i. It could be cool if you do so, to avoid to add 7 commits into PrestaShop main history! :D

Thanks for the tips @boherm!
I followed the guide you linked to me and after several attempts I was able to clear the history.
I hope I did everything correctly.
Thanks again!

**Review by boherm — DISMISSED** [review]
LGTM :)
Thanks @Codencode!

**Comment by Hlavtox** [comment]
Now I am thinking.... The address validation could fail for many reasons. Since we have no way to tell object model not to die when something goes wrong.

What about doing this instead?
```php
try {
    $fields = $address->getFields();
} catch (Exception $e) {
    return sha1('Invalid address');
}
```

**Comment by Codencode** [comment]
> Now I am thinking.... The address validation could fail for many reasons. Since we have no way to tell object model not to die when something goes wrong.
> 
> What about doing this instead?
> 
> ```
> try {
>     $fields = $address->getFields();
> } catch (Exception $e) {
>     return sha1('Invalid address');
> }
> ```
Indeed what you say is correct and your solution covers all cases.

@boherm what do you think?

**Comment by boherm** [comment]
@Hlavtox, @Codencode, Yes, I agree with you guys.
With this way all cases can be covered and I think that is the best solution to fix this issue too.

**Comment by Codencode** [comment]
> @Hlavtox, @Codencode, Yes, I agree with you guys. With this way all cases can be covered and I think that is the best solution to fix this issue too.

Okay,
then I proceed with the changes.
Thank you!

**Comment by Codencode** [comment]
@boherm, @Hlavtox, I made the change, but **PHPUnit (7.4)** failed and I don't understand why.

Can you confirm that the modification is ok?

Thanks!

**Comment by boherm** [comment]
@Codencode, Github Actions may fail sometime for obscur reasons. :-D 
It seems to be the case for `PHP / PHPUnit (7.4) (pull_request)` and I think if we relaunch this check it could be run with success.

**Comment by Codencode** [comment]
> @Codencode, Github Actions may fail sometime for obscur reasons. :-D
> It seems to be the case for `PHP / PHPUnit (7.4) (pull_request)` and I think if we relaunch this check it could be run with success.

Ok, if I wanted to relaunch only this control how should I do?

**Comment by boherm** [comment]
Good question, but I ask to restart this job and It's green! :-D
(thanks @jolelievre)

**Review by boherm — DISMISSED** [review]

**Review by Hlavtox — APPROVED** [review]
Looks good, did not test it, but I think we found out a pretty neat solution guys!

**Review by Hlavtox — CHANGES_REQUESTED** [review]
Just remove the duplicate $uniqId = ''; :-)

**Comment by Codencode** [comment]
> Just remove the duplicate $uniqId = ''; :-)

I hadn't really noticed the duplicate variable.
Do I rebase? so we have clean code?

Thank you!

**Comment by Codencode** [comment]
> Just remove the duplicate $uniqId = ''; :-)

I hadn't noticed the duplicate variable.
I made the change and did the rebase so I have clean code, I hope I did it right.

Thanks @Hlavtox!

**Review by boherm — APPROVED** [review]

**Review by Hlavtox — APPROVED** [review]

**Comment by Codencode** [comment]
@Hlavtox, @boherm, 
do i have to do anything else?

**Comment by Hlavtox** [comment]
Ping @PrestaShop/committers

**Review by kpodemski — APPROVED** [review]

**Review by djoelleuch — APPROVED** [review]
Hello @Codencode  ,
I have checked your PR the issue is well fixed ,

[32290.webm](https://user-images.githubusercontent.com/115152460/236502221-f1ce9510-78f8-43a3-ace7-3ace5b10508b.webm)

It's QA approved :heavy_check_mark:
Thank you for you PR :rocket:

**Comment by nicosomb** [comment]
Thank you @Codencode !

**Pull request merged**
